MAINT validate parameter in AdaBoostClassifier and AdaBoostRegressor

berkecanrizai opened the pull request:
I have removed unnecessary checks under the fit method.

I will be adjusting the test files as well. 
See;
[#### Reference Issues/PRs](https://github.com/scikit-learn/scikit-learn/issues/23462)



Adds new validation method for parameters from #23462 for AdaBoostClassifier and AdaBoostRegressor.


I am not totally sure if the check for base_estimator is correctly implemented currently, would love feedback on that. Thanks.


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

berkecanrizai committed:
Added new validation method for adaboostclassifier and adaregressor. Removed unnecessary check in fit methods.

SandroEph reviewed (commented):
As `AdaBoostClassifier` and `AdaBoostRegressor` both inherit from the base class `BaseWeightBoosting`, in my opinion it would be better to refactor the parameter validation inside the parent class and add the different parameters in the respective classes (eg. `algorithm` for `AdaBoostClassifier` and `loss` for `AdaBoostRegressor`)

glemaitre reviewed (commented):
Indeed you should use properly the inheritance here.

I attach a patch showing the diff required to make the change:

```diff
diff --git a/sklearn/ensemble/_weight_boosting.py b/sklearn/ensemble/_weight_boosting.py
index cb3116148f..d66dcf160a 100644
--- a/sklearn/ensemble/_weight_boosting.py
+++ b/sklearn/ensemble/_weight_boosting.py
@@ -33,7 +33,13 @@ import warnings
 from scipy.special import xlogy
 
 from ._base import BaseEnsemble
-from ..base import ClassifierMixin, RegressorMixin, is_classifier, is_regressor
+from ..base import (
+    BaseEstimator,
+    ClassifierMixin,
+    RegressorMixin,
+    is_classifier,
+    is_regressor,
+)
 
 from ..tree import DecisionTreeClassifier, DecisionTreeRegressor
 from ..utils import check_random_state, _safe_indexing
@@ -45,10 +51,7 @@ from ..utils.validation import check_is_fitted
 from ..utils.validation import _check_sample_weight
 from ..utils.validation import has_fit_parameter
 from ..utils.validation import _num_samples
-from ..utils._param_validation import StrOptions
-from ..utils._param_validation import Interval
-from numbers import Integral
-from numbers import Real
+from ..utils._param_validation import Interval, StrOptions
 
 __all__ = [
     "AdaBoostClassifier",
@@ -63,6 +66,13 @@ class BaseWeightBoosting(BaseEnsemble, metaclass=ABCMeta):
     instead.
     """
 
+    _parameter_constraints = {
+        "base_estimator": [None, BaseEstimator],
+        "n_estimators": [Interval(numbers.Integral, 1, None, closed="left")],
+        "learning_rate": [Interval(numbers.Real, 0, None, closed="neither")],
+        "random_state": ["random_state"],
+    }
+
     @abstractmethod
     def __init__(
         self,
@@ -115,23 +125,7 @@ class BaseWeightBoosting(BaseEnsemble, metaclass=ABCMeta):
         -------
         self : object
         """
-        # Validate scalar parameters
-        check_scalar(
-            self.n_estimators,
-            "n_estimators",
-            target_type=numbers.Integral,
-            min_val=1,
-            include_boundaries="left",
-        )
-
-        check_scalar(
-            self.learning_rate,
-            "learning_rate",
-            target_type=numbers.Real,
-            min_val=0,
-            include_boundaries="neither",
-        )
-
+        self._validate_params()
         X, y = self._validate_data(
             X,
             y,
@@ -460,12 +454,8 @@ class AdaBoostClassifier(ClassifierMixin, BaseWeightBoosting):
     """
 
     _parameter_constraints = {
-        "base_estimator": [None, "base_estimator"],
-        "n_estimators": [Interval(Integral, 1, None, closed="left")],
-        "learning_rate": [Interval(Real, 0, None, closed="neither")],
-        "loalgorithmss": [StrOptions({"SAMME", "SAMME.R"})],
-        "loss": [StrOptions({"linear", "square", "exponential"})],
-        "random_state": ["random_state"],
+        **BaseWeightBoosting._parameter_constraints,
+        "algorithm": [StrOptions({"SAMME", "SAMME.R"})],
     }
 
     def __init__(
@@ -487,31 +477,6 @@ class AdaBoostClassifier(ClassifierMixin, BaseWeightBoosting):
 
         self.algorithm = algorithm
 
-    def fit(self, X, y, sample_weight=None):
-        """Build a boosted classifier from the training set (X, y).
-
-        Parameters
-        ----------
-        X : {array-like, sparse matrix} of shape (n_samples, n_features)
-            The training input samples. Sparse matrix can be CSC, CSR, COO,
-            DOK, or LIL. COO, DOK, and LIL are converted to CSR.
-
-        y : array-like of shape (n_samples,)
-            The target values (class labels).
-
-        sample_weight : array-like of shape (n_samples,), default=None
-            Sample weights. If None, the sample weights are initialized to
-            ``1 / n_samples``.
-
-        Returns
-        -------
-        self : object
-            Fitted estimator.
-        """
-        self._validate_params()
-
-        # Fit
-        return super().fit(X, y, sample_weight)
 
     def _validate_estimator(self):
         """Check the estimator and set the base_estimator_ attribute."""
@@ -1047,11 +1012,8 @@ class AdaBoostRegressor(RegressorMixin, BaseWeightBoosting):
     """
 
     _parameter_constraints = {
-        "base_estimator": [None, "base_estimator"],
-        "n_estimators": [Interval(Integral, 1, None, closed="left")],
-        "learning_rate": [Interval(Real, 0, None, closed="neither")],
+        **BaseWeightBoosting._parameter_constraints,
         "loss": [StrOptions({"linear", "square", "exponential"})],
-        "random_state": ["random_state"],
     }
 
     def __init__(
@@ -1074,31 +1036,6 @@ class AdaBoostRegressor(RegressorMixin, BaseWeightBoosting):
         self.loss = loss
         self.random_state = random_state
 
-    def fit(self, X, y, sample_weight=None):
-        """Build a boosted regressor from the training set (X, y).
-
-        Parameters
-        ----------
-        X : {array-like, sparse matrix} of shape (n_samples, n_features)
-            The training input samples. Sparse matrix can be CSC, CSR, COO,
-            DOK, or LIL. COO, DOK, and LIL are converted to CSR.
-
-        y : array-like of shape (n_samples,)
-            The target values (real numbers).
-
-        sample_weight : array-like of shape (n_samples,), default=None
-            Sample weights. If None, the sample weights are initialized to
-            1 / n_samples.
-
-        Returns
-        -------
-        self : object
-            Fitted AdaBoostRegressor estimator.
-        """
-        self._validate_params()
-
-        # Fit
-        return super().fit(X, y, sample_weight)
 
     def _validate_estimator(self):
         """Check the estimator and set the base_estimator_ attribute."""
diff --git a/sklearn/tests/test_common.py b/sklearn/tests/test_common.py
index 920a1c2bc8..fcaf71edcf 100644
--- a/sklearn/tests/test_common.py
+++ b/sklearn/tests/test_common.py
@@ -447,8 +447,6 @@ def test_estimators_do_not_raise_errors_in_init_or_set_params(Estimator):
 
 PARAM_VALIDATION_ESTIMATORS_TO_IGNORE = [
     "ARDRegression",
-    "AdaBoostClassifier",
-    "AdaBoostRegressor",
     "AdditiveChi2Sampler",
     "AffinityPropagation",
     "AgglomerativeClustering",
```

glemaitre committed:
Merge branch 'main' into val_param_AdaBoost_regressor

berkecanrizai committed:
Removed unnecessary test and modified another one, test_adaboost_classifier_sample_weight_error.

berkecanrizai committed:
Inherited _parameter_constraints from BaseWeightBoosting for AdaBoostclassifier and regressor. Removed overriden fit from these classes as they inherit from super without modification. Removed one unnecessary import.

berkecanrizai commented:
@glemaitre @SandroEph thanks for the recommendations. I have added suggested changes and removed one unnecessary import "check_scalars". It should be fine currently. Let me know if anything needs to be changed.

glemaitre committed:
remove useless parametrization

glemaitre committed:
Merge branch 'main' into val_param_AdaBoost_regressor

glemaitre commented:
You forgot to remove the parametrization together with the test.
I just make a small push such that we can run the CIs again with this fix.

It should be OK now.

glemaitre reviewed (approved):
LGTM

jeremiedbb committed:
remove estimators from common test blacklist

jeremiedbb committed:
Merge remote-tracking branch 'upstream/main' into pr/berkecanrizai/23773

jeremiedbb committed:
lint

jeremiedbb committed:
mypy ignore error

jeremiedbb committed:
numpydoc

jeremiedbb reviewed (approved):
LGTM. Thanks @berkecanrizai
